core/src/data_sources/splitter.rs

```diff
@@ -1,13 +1,379 @@
+/// `max_chunk_size` limit (see `fn chunks`).
+#[derive(Debug, Clone)]
+pub struct TokenizedSection {
+    pub max_chunk_size: usize,
+    pub prefixes: Vec<TokenizedText>,
+    pub tokens_count: usize,
+    pub content: Option<TokenizedText>,
+    pub sections: Vec<TokenizedSection>,
+}
+
+impl TokenizedSection {
+    #[async_recursion]
+    pub async fn from(
+        embedder: &Box<dyn Embedder + Sync + Send>,
+        max_chunk_size: usize,
+        mut prefixes: Vec<TokenizedText>,
+        section: &Section,
+    ) -> Result<Self> {
+        let (prefix, mut content) = try_join!(
+            TokenizedText::from(embedder, section.prefix.as_ref()),
+            TokenizedText::from(embedder, section.content.as_ref())
+        )?;
+
+        // Add the new prefix to the list of prefixes to be passed down children.
+        match prefix.as_ref() {
+            Some(prefix) => {
+                prefixes.push(prefix.clone());
+            }
+            None => (),
+        };
+
+        let prefixes_tokens_count = prefixes.iter().map(|p| p.tokens.len()).sum::<usize>();
+        if prefixes_tokens_count >= max_chunk_size / 2 {
+            Err(anyhow!(
+                "Could not tokenize the provided document,
+                 prefixes accrue to more than half `max_chunk_size`"
+            ))?;
+        }
+
+        let mut sections: Vec<TokenizedSection> = vec![];
+
+        // Create new children for content if content overflows max_chunk_size. This will result in
+        // moving the content in childrens that will be leaf nodes.
+        if let Some(c) = content.as_ref() {
+            let effective_max_chunk_size = max_chunk_size - prefixes_tokens_count;
+
+            if c.tokens.len() > effective_max_chunk_size {
+                let splits = split_text(&embedder, effective_max_chunk_size, &c.text).await?;
+
+                // Prepend to childrens the splits of the content with no additional prefixes (they
+                // will inherit the current section prefixes whose content will be removed).
+                sections.extend(
+                    splits
+                        .into_iter()
+                        .map(|t| TokenizedSection {
+                            max_chunk_size,
+                            prefixes: prefixes.clone(),
+                            tokens_count: prefixes_tokens_count + t.tokens.len(),
+                            content: Some(t),
+                            sections: vec![],
+                        })
+                        .collect::<Vec<_>>(),
+                );
+
+                // Remove the content from the current section.
+                content = None;
+            }
+        }
+
+        // Create a new leaf children for content if prefix is not None and content is not None to
+        // preserve the invariant that content nodes are leaf nodes.
+        if let Some(c) = content.as_ref() {
```

Comment:
what about merging that with the code above (if prefix, splits = singleton array)?